Sources/NIOCore/ByteBuffer-hexdump.swift

```diff
@@ -0,0 +1,188 @@
+
+extension ByteBuffer {
+
+    /// Describes a ByteBuffer hexDump format.
+    /// Can be either xxd output compatible, or hexdump compatible.
+    public struct HexDumpFormat {
+        enum Value {
+            case xxdCompatible(maxLength: Int? = nil)
+            case hexDumpCompatible(maxLength: Int? = nil)
+        }
+
+        var value: Value
+        init(_ value: Value) { self.value = value }
+
+        /// A hex dump format compatible with `xxd` CLI utility.
+        public static let xxdCompatible = Self(.xxdCompatible(maxLength: nil))
+
+        /// A hex dump format compatible with `hexdump` command line utility.
+        public static let hexDumpCompatible = Self(.hexDumpCompatible(maxLength: nil))
+
+        /// A hex dump format compatible with `xxd`, clipped to `maxLength` bytes dumped.
+        /// This format will dump first maxLength / 2 bytes, and the last maxLength / 2 bytes, replacing the rest with " ... ".
+        public static func xxdCompatible(maxLength: Int) -> Self {
+            Self(.xxdCompatible(maxLength: maxLength))
+        }
+
+        /// A hex dump format compatible with `hexdump` command line tool.
+        /// This format will dump first maxLength / 2 bytes, and the last maxLength / 2 bytes, with a placeholder in between.
+        public static func hexDumpCompatible(maxLength: Int) -> Self {
+            Self(.hexDumpCompatible(maxLength: maxLength))
+        }
+    }
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer,
+    /// in a format that's compatible with `xxd -r -p`.
+    /// `hexDumpShort()` always dumps all readable bytes, i.e. from `readerIndex` to `writerIndex`,
+    /// so you should set those indices to desired location to get the offset and length that you need to dump.
+    func hexDumpShort() -> String {
+        var hexString = ""
+        hexString.reserveCapacity(self.readableBytes * 3)
+
+        for byte in self.readableBytesView {
+            hexString += String(byte: byte, padding: 2)
+            hexString += " "
+        }
+
+        return String(hexString.dropLast())
+    }
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer.,
+    /// in a format that's compatible with `xxd -r -p`, but clips the output to the max length of `limit` bytes.
+    /// If the dump contains more than the `limit` bytes, this function will return the first `limit/2`
+    /// and the last `limit/2` of that, replacing the rest with `...`, i.e. `01 02 03 ... 09 11 12`.
+    ///
+    /// - parameters:
+    ///     - limit: The maximum amount of bytes presented in the dump.
+    func hexDumpShort(limit: Int) -> String {
+        guard self.readableBytes > limit else {
+            return self.hexDumpShort()
+        }
+
+        var buffer = self
+        let front = buffer.readSlice(length: limit / 2)!
+        buffer.moveReaderIndex(forwardBy: buffer.readableBytes - limit / 2)
```

Comment:
It might be cleaner to use `moveReaderIndex(to:)` with a computation from `writerIndex`. Up to you though.